[SPARK-29779][CORE] Compact old event log files and cleanup

HeartSaVioR opened the pull request:
### What changes were proposed in this pull request?

This patch proposes to compact old event log files when end users enable rolling event log, and clean up these files after compaction. 

Here the "compaction" really mean is filtering out listener events for finished/removed things - like jobs which take most of space for event log file except SQL related events. To achieve this, compactor does two phases reading: 1) tracking the live jobs (and more to add) 2) filtering events via leveraging the information about live things and rewriting to the "compacted" file.

This approach retains the ability of compatibility on event log file and adds the possibility of reducing the overall size of event logs. There's a downside here as well: executor metrics for tasks would be inaccurate, as compactor will filter out the task events which job is finished, but I don't feel it as a blocker.

Please note that SPARK-29779 leaves below functionalities for future JIRA issue as the patch for SPARK-29779 is too huge and we decided to break down:

* apply filter in SQL events
* integrate compaction into FsHistoryProvider
* documentation about new configuration

### Why are the changes needed?

One of major goal of SPARK-28594 is to prevent the event logs to become too huge, and SPARK-29779 achieves the goal. We've got another approach in prior, but the old approach required models in both KVStore and live entities to guarantee compatibility, while they're not designed to do so.

### Does this PR introduce any user-facing change?

No.

### How was this patch tested?

Added UTs.

HeartSaVioR committed:
[SPARK-29779][CORE] Compact old event log files and cleanup - part 1

HeartSaVioR commented:
Please note that this is a part of #26416 - related review comments are addressed.

cc. @vanzin @gaborgsomogyi @Ngone51 as reviewers of #26416.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/20860/
Test PASSed.

SparkQA commented:
**[Test build #116068 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/116068/testReport)** for PR 27085 at commit [`5fd8f27`](https://github.com/apache/spark/commit/5fd8f27e6289de0a23f19f951533a5f89a7dff5c).

SparkQA commented:
**[Test build #116068 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/116068/testReport)** for PR 27085 at commit [`5fd8f27`](https://github.com/apache/spark/commit/5fd8f27e6289de0a23f19f951533a5f89a7dff5c).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `  case class FilterStatistics(`
  * `class EventLogFileCompactor(`
  * `class FilteredEventLogFileRewriter(`
  * `class CompactedEventLogFileWriter(`

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/116068/
Test FAILed.

SparkQA commented:
**[Test build #116075 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/116075/testReport)** for PR 27085 at commit [`704c6ca`](https://github.com/apache/spark/commit/704c6ca3c3db7b02c38edbe717df46fc63b1d3e4).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/20867/
Test PASSed.

HeartSaVioR committed:
Address events out of order

SparkQA commented:
**[Test build #116078 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/116078/testReport)** for PR 27085 at commit [`c06575c`](https://github.com/apache/spark/commit/c06575c8843e3655f32e391782f8ddb9d8899c2f).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/20870/
Test PASSed.

SparkQA commented:
**[Test build #116078 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/116078/testReport)** for PR 27085 at commit [`c06575c`](https://github.com/apache/spark/commit/c06575c8843e3655f32e391782f8ddb9d8899c2f).
 * This patch **fails due to an unknown error code, -9**.
 * This patch merges cleanly.
 * This patch adds no public classes.

SparkQA commented:
**[Test build #116075 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/116075/testReport)** for PR 27085 at commit [`704c6ca`](https://github.com/apache/spark/commit/704c6ca3c3db7b02c38edbe717df46fc63b1d3e4).
 * This patch **fails due to an unknown error code, -9**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `  case class FilterStatistics(`
  * `class EventLogFileCompactor(`
  * `class FilteredEventLogFileRewriter(`
  * `class CompactedEventLogFileWriter(`

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/116078/
Test FAILed.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/116075/
Test FAILed.

HeartSaVioR commented:
Hmm... it seems hard to filter events accurately, assuming arbitrary out of order for events.

If there's no idea to make it accurate, we may want to choose the safer approach which may filter out less events on out of order, as filtered out events will be lost once we compact and there's no way to restore them back. It may leave some unmatched events which might have been filtered out if they are in order, but (SQL) AppStatusListener have been tolerating events out of order so it seems acceptable.

HeartSaVioR committed:
safer condition for filtering out events to deal with events out of order

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/20880/
Test PASSed.

SparkQA commented:
**[Test build #116088 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/116088/testReport)** for PR 27085 at commit [`f92b0cb`](https://github.com/apache/spark/commit/f92b0cbe40c0f198c82bcea07830da668da9eacf).

SparkQA commented:
**[Test build #116088 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/116088/testReport)** for PR 27085 at commit [`f92b0cb`](https://github.com/apache/spark/commit/f92b0cbe40c0f198c82bcea07830da668da9eacf).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/116088/
Test PASSed.

vanzin commented on core/src/main/scala/org/apache/spark/deploy/history/EventLogFileCompactor.scala:
```diff
@@ -0,0 +1,212 @@
+    } else {
+      val builders = initializeBuilders(fs, filesToCompact.map(_.getPath))
+
+      val filters = builders.map(_.createFilter())
```
minor: can avoid this if score is too low.

HeartSaVioR replied:
The score is calculated under the information of EventFilter despite the fact the information is actually built from EventFilterBuilder. Technically there will be no difference if we add EventFilterBuilder to provide statistics, but it's semantically different.

The difference of two interfaces is that EventFilterBuilder is in inconsistent state, as flowing event into listener bus will change the state of builder, whereas EventFilter is a snapshot holding the state for loaded; so the event filter builder should be only used to build the event filter, and event filter will be used for all other things.

vanzin replied:
Nevermind, I missed this was used in the very next line.

vanzin commented on core/src/main/scala/org/apache/spark/deploy/history/EventLogFileCompactor.scala:
```diff
@@ -0,0 +1,212 @@
+ * [[SingleEventLogFileWriter]], but only `originalFilePath` is used to determine the
+ * path of compact file.
+ */
+class CompactedEventLogFileWriter(
```
private

vanzin replied:
private

vanzin commented on core/src/main/scala/org/apache/spark/deploy/history/EventLogFileCompactor.scala:
```diff
@@ -0,0 +1,212 @@
+ * decide to reject the event or don't mind about the event. Otherwise, the original line for
+ * the event is written to the compact file as it is.
+ */
+class FilteredEventLogFileRewriter(
```
private. If fact, this class doesn't really have any state, can it just be a method in `EventLogFileCompactor`?

HeartSaVioR replied:
OK will include to compactor.

vanzin commented on core/src/main/scala/org/apache/spark/deploy/history/EventLogFileCompactor.scala:
```diff
@@ -0,0 +1,212 @@
+ * decide to reject the event or don't mind about the event. Otherwise, the original line for
+ * the event is written to the compact file as it is.
+ */
+class FilteredEventLogFileRewriter(
```
Can be private. In fact, this class has no state, can it just be a method in `EventLogFileCompactor`? You can leave the method public so you can test it directly.

vanzin commented on core/src/main/scala/org/apache/spark/internal/config/package.scala:
```diff
@@ -195,6 +195,24 @@ package object config {
+        s"and ${EVENT_LOG_ROLLING_MAX_FILE_SIZE.key} accordingly if you want to control " +
+        "the overall size of event log files.")
+      .intConf
+      .checkValue(_ > 0, "Max event log files to retain should be higher than 0.")
```
Couldn't it be 0, potentially? So you only keep one last compacted log file, instead of one compacted + one normal.

HeartSaVioR replied:
If I'm not missing, it can't be, as setting this 0 means we allow compacting the event log file which running app is writing now. It also doesn't make sense for finished app as well, since all jobs should have been finished when app is finished - so compaction would lose all job-related events.

vanzin replied:
Hmm, I see. Merging the last two logs might still result in some savings, but probably not worth it if it means having extra logic to say "don't filter out things that started in the last log file".

On a side note, compaction currently overrides all the "retained blah" configurations; so at the end you may end up, for example, with very few jobs in the Spark UI, whereas the retention settings say you should keep the last 1000 jobs or so.

HeartSaVioR replied:
> On a side note, compaction currently overrides all the "retained blah" configurations

Yeah, that's actually the point I have been thinking about, as end users may see only few of jobs if they set this to 1 and the log files are just compacted.

Maybe we could improve event filter to configure max retained jobs on compact file so that some of finished jobs can be still tracked (`live jobs + finished jobs <= max retained jobs`). It cannot be strictly max, as live jobs could have been more than the configured value, so may need to have better name.

Looks like it could be further TODO worth filing JIRA issue.

vanzin commented on core/src/test/scala/org/apache/spark/deploy/history/BasicEventFilterBuilderSuite.scala:
```diff
@@ -0,0 +1,233 @@
+    val s0Metrics = TaskMetrics.empty
+    s0Metrics.setExecutorCpuTime(2L)
+    s0Metrics.setExecutorRunTime(4L)
+
```
OOC, is this important for the test? Seems random.

HeartSaVioR replied:
Ah yes, it's actually copied from other test, as the original test seems to deal with complicated case and I wanted to be sure the builder handles the case correctly. But it might be unnecessarily complicated in point of testing the functionality of BasicEventFilterBuilder.

Would we be better to simplify the case, or just need to remove some lines?

vanzin commented on core/src/test/scala/org/apache/spark/deploy/history/BasicEventFilterSuite.scala:
```diff
@@ -0,0 +1,199 @@
+    val liveTasks = Set(3L, 4L, 5L, 6L)
+    val liveRDDs = Set(3, 4, 5, 6)
+    val liveExecutors: Set[String] = Set("1", "2")
+    val filterStats = FilterStatistics(2, 1, 2, 1, 4, 2)
```
instead of magic numbers, use e.g. `liveJobs.size`?

vanzin commented on core/src/test/scala/org/apache/spark/deploy/history/BasicEventFilterSuite.scala:
```diff
@@ -0,0 +1,199 @@
+    assert(acceptFn(createExecutorAddedEvent(1)) === Some(false))
+    // though the name of event is stage executor metrics, AppStatusListener only deals with
+    // live executors
+    assert(acceptFn(
```
nit: this makes it look like `Some(false)` is an argument to `acceptFn`.

HeartSaVioR replied:
Good point. I'll swap the expected value and the actual call.

vanzin commented on core/src/test/scala/org/apache/spark/deploy/history/BasicEventFilterSuite.scala:
```diff
@@ -0,0 +1,199 @@
+  }
+
+  test("other events should be left to other filters") {
+    def assertNone(predicate: => Option[Boolean]): Unit = {
```
This may look like a good idea but it makes `assert` generate generic, non-helpful messages.

HeartSaVioR replied:
Yeah agreed. Thought too much on reducing the code. Good point.

vanzin commented on core/src/test/scala/org/apache/spark/deploy/history/BasicEventFilterSuite.scala:
```diff
@@ -0,0 +1,199 @@
+    assertNone(acceptFn(SparkListenerApplicationEnd(1)))
+    val bmId = BlockManagerId("1", "host1", 1)
+    assertNone(acceptFn(SparkListenerBlockManagerAdded(0, bmId, 1)))
+    assertNone(acceptFn(SparkListenerBlockManagerRemoved(1, bmId)))
```
See previous comment about these events.

vanzin commented on core/src/test/scala/org/apache/spark/deploy/history/BasicEventFilterSuite.scala:
```diff
@@ -0,0 +1,199 @@
+    assertNone(acceptFn(SparkListenerBlockManagerAdded(0, bmId, 1)))
+    assertNone(acceptFn(SparkListenerBlockManagerRemoved(1, bmId)))
+    assertNone(acceptFn(SparkListenerBlockUpdated(
+      storage.BlockUpdatedInfo(bmId, RDDBlockId(1, 1), StorageLevel.DISK_ONLY, 0, 10))))
```
Ah, maybe this one also needs to be filtered if the executor is gone?

(Probably not that important since these events are not in the event logs by default.)

HeartSaVioR replied:
For now the code doesn't track the relation between block manager and executor; we may need to add more info if we would like to apply filter to SparkListenerBlockUpdated events. Would we like to do it, or leave as it is?

vanzin replied:
For the UI, executor and block manager are the same thing. Everything is tracked as "the executor".

As I mentioned before, if you don't filter out block manager events, then you're basically defeating any filtering of executor-related events, since block manager events make AppStatusListener start tracking the executor:

```
  override def onBlockManagerAdded(event: SparkListenerBlockManagerAdded): Unit = {
    // This needs to set fields that are already set by onExecutorAdded because the driver is
    // considered an "executor" in the UI, but does not have a SparkListenerExecutorAdded event.
    val exec = getOrCreateExecutor(event.blockManagerId.executorId, event.time)
```

So as far as the data that shows up in the UI, you'll get the executor there, maybe with some missing information.

Focusing on this particular even is probably moot, because it's not written to event logs; the history UI currently doesn't have any RDD storage information. But it would be good to understand how the UI behaves w.r.t. executor information after you apply the filters.

vanzin commented on core/src/test/scala/org/apache/spark/deploy/history/BasicEventFilterSuite.scala:
```diff
@@ -0,0 +1,199 @@
+    assertNone(acceptFn(SparkListenerLogStart("testVersion")))
+  }
+
+  private def assertFilterJobEvents(
```
This is called in a single place. Inline.

vanzin commented on core/src/test/scala/org/apache/spark/deploy/history/EventLogFileCompactorSuite.scala:
```diff
@@ -0,0 +1,233 @@
+      originalFiles: Seq[FileStatus],
+      compactRet: (CompactionResult.Value, Option[Long]),
+      expectedCompactRet: CompactionResult.Value): Unit = {
+  assert(compactRet._1 === expectedCompactRet)
```
indentation

vanzin commented on core/src/test/scala/org/apache/spark/deploy/history/FilteredEventLogFileRewriterSuite.scala:
```diff
@@ -0,0 +1,90 @@
+
+      writer.stop()
+
+      val filters = Seq(new TestEventFilter1, new TestEventFilter2)
```
These filter are only used here (even in your other PRs, AFAICT). It would be easier to follow if they were also defined here, since the comments above explaining the expectation reflect the behavior of these filters.

HeartSaVioR replied:
Good finding. I might have been used for other tests as well but removed afterwards. Will move.

vanzin commented on core/src/test/scala/org/apache/spark/deploy/history/FilteredEventLogFileRewriterSuite.scala:
```diff
@@ -0,0 +1,90 @@
+          linesLength += 1
+          assert(expectedLines.contains(line))
+        }
+        assert(linesLength === expectedLines.length, "The number of lines for rewritten file " +
```
The assert already generates pretty much that same error message.

vanzin commented on core/src/test/scala/org/apache/spark/status/ListenerEventsTestHelper.scala:
```diff
@@ -0,0 +1,201 @@
+      stageToTaskIds: Map[Int, Seq[Long]],
+      stageToRddIds: Map[Int, Seq[Int]])
+
+  def pushJobEventsWithoutJobEnd(
```
Seems like this is only used in a later PR, might be good to add it there.

HeartSaVioR replied:
OK will remove it here and restore in later PR.

vanzin reviewed (commented):
Still a lot of code, but more manageable.

vanzin commented:
Also I'd reword the PR title a little, especially in the subsequent PRs.

HeartSaVioR commented on core/src/main/scala/org/apache/spark/deploy/history/BasicEventFilterBuilder.scala:
```diff
@@ -0,0 +1,177 @@
+  override def onStageSubmitted(stageSubmitted: SparkListenerStageSubmitted): Unit = {
+    val stageId = stageSubmitted.stageInfo.stageId
+    if (_stageToRDDs.get(stageId).isEmpty) {
+      // job start event is not received yet
```
I admit it's just guessing that events out of order can happen in arbitrary way, as previous review comment pointed out I should consider events out of order but I have no idea about these out of order. But it's also non-trivial to sort out which cases can happen from looking into AppStatusListener code.

Why not enumerating the cases we've observed in AppStatusListener as code comment? There's no value if only part of group knows about it in their memory. This looks to be a "truck number".

Btw, have we struggle to find out why these events can be out of order, and try to fix it?

vanzin replied:
The order of job/stage/task events is defined by the DAGScheduler. Adding comments in `AppStatusListener` would just be trying to document the scheduler behavior, so if you want to track that, it would be better to document the scheduler itself. When you add other subsystems (like SQL), then you also add other things that may arrive out of order (see how the SQL listener needs to handle job / execution events).

On the scheduler side, I think most of the out of order issues are related to tasks. I'm somewhat confident that job and stage events generally arrive in order, but tasks can be very off. e.g. see this comment in `handleBeginEvent`:

```
    // Note that there is a chance that this task is launched after the stage is cancelled.
    // In that case, we wouldn't have the stage anymore in stageIdToStage.
```

I guess my reply here applies more to my `onTaskStart` comment below; here we can assume that the events are in order. But there, it would be better to ignore events if they arrive at the wrong time.

HeartSaVioR replied:
Ah OK my bad. I was confused that there's a case the message bus doesn't guarantee event order (so I felt we have to rely on history), but looks like that's not the case. Then it would be enough if there's enough comment on DAGScheduler. 

Thanks for the correction. I may need to spend my time to read through DAGScheduler, but collecting the information from your comment, looks like the main point is discarding the task if the stage is not active. I'll deal with it for now.

HeartSaVioR commented on core/src/main/scala/org/apache/spark/deploy/history/BasicEventFilterBuilder.scala:
```diff
@@ -0,0 +1,177 @@
+
+  override def onTaskStart(taskStart: SparkListenerTaskStart): Unit = {
+    totalTasks += 1
+    val curTasks = _stageToTasks.getOrElseUpdate(taskStart.stageId,
```
Same here; this is assuming arbitrary out of order of events. If we are sure stage submission event must happen earlier than task start event, this code doesn't make sense. But I also would like to be sure we document these cases as a reference. It's not easy to understand the intention if I don't have background knowledge and can't find it easily.

HeartSaVioR commented on core/src/main/scala/org/apache/spark/deploy/history/BasicEventFilterBuilder.scala:
```diff
@@ -0,0 +1,177 @@
+    case e: SparkListenerUnpersistRDD =>
+      liveRDDs.contains(e.rddId)
+    case e: SparkListenerExecutorMetricsUpdate =>
+      e.accumUpdates.exists { case (_, stageId, _, _) => liveStages.contains(stageId) }
```
In AppStatusListener, SparkListenerExecutorMetricsUpdate updates metrics not only for executor but also for stage and task; so ideally the event may need to be accepted if either task or stage is live. (I'll add the check for liveTasks as well)

If checking executor is needed in any way, we may want to move this to BasicEventFilter.

vanzin replied:
Hmm, ok, makes sense to try to keep the stage metrics accurate.

HeartSaVioR commented on core/src/main/scala/org/apache/spark/deploy/history/BasicEventFilterBuilder.scala:
```diff
@@ -0,0 +1,177 @@
+    case e: SparkListenerExecutorRemoved => liveExecutors.contains(e.executorId)
+    case e: SparkListenerExecutorBlacklisted => liveExecutors.contains(e.executorId)
+    case e: SparkListenerExecutorUnblacklisted => liveExecutors.contains(e.executorId)
+    case e: SparkListenerStageExecutorMetrics => liveExecutors.contains(e.execId)
```
The intention is not rejecting these kinds of events as any implementations of EventFilter don't know how to determine whether it should be accepted or not, but if we have some cases whether it should never be rejected, it would be better to explicitly accept these cases for safety.

Thanks for guiding on the case as driver being a block manager. Looking into the code of AppStatusListener, there seems to be no information to determine whether the block manager is driver or not; it seems safer to accept all.

vanzin replied:
We shouldn't be accepting block manager events for executors that are dead. That basically means that even after you compact the log files, you'll get all the executors in the UI. It defeats the filtering of the executor added / removed events.

The driver is different because it does not generate an executor added event, just a block manager added event. So when filtering block manager events you need to check whether the block manager ID refers either to the driver or a live executor.

HeartSaVioR commented on core/src/main/scala/org/apache/spark/deploy/history/EventLogFileCompactor.scala:
```diff
@@ -0,0 +1,212 @@
+ *
+ * This class assumes caller will provide the sorted list of files which are sorted by the index of
+ * event log file, with "at most" one compact file placed first if it exists. Caller should keep in
+ * mind that this class doesn't care about the semantic of ordering.
```
Yeah, it may need to be reworded, or maybe just removed as looks like previous sentence already explains the requirement for caller. 

I'll remove the sentence for now; please let me know if the sentence doesn't seem sufficient to represent the requirement.

HeartSaVioR committed:
Reflect review comments

SparkQA commented:
**[Test build #116126 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/116126/testReport)** for PR 27085 at commit [`9a7a210`](https://github.com/apache/spark/commit/9a7a210d8b1214d7d9251bcf6b32a412e1d6e803).

HeartSaVioR commented:
Thanks for the detailed reviews, as always! I've addressed review comments, or left comments. Please take a look again. Thanks in advance!
(Happy new year, btw. :) )

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/20919/
Test PASSed.

HeartSaVioR commented:
Regarding the title, it seems OK to drop "part" as follow-up PRs could have another meaningful title; this PR seems to be better to be original title, although this PR doesn't integrate the new feature into SHS.

Btw, would it be better to file a new issue for each follow-up PR, or OK to put all of them into SPARK-29779?

SparkQA commented:
**[Test build #116126 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/116126/testReport)** for PR 27085 at commit [`9a7a210`](https://github.com/apache/spark/commit/9a7a210d8b1214d7d9251bcf6b32a412e1d6e803).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `case class CompactionResult(code: CompactionResultCode.Value, compactIndex: Option[Long])`

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/116126/
Test FAILed.

HeartSaVioR commented:
Test failure on build 116126 is below:

```
org.apache.spark.scheduler.TaskSetManagerSuite.reset
 Error Details
org.scalatest.exceptions.TestFailedException: task0.isDefined was true, but task1.isDefined was false
 Stack Trace
sbt.ForkMain$ForkError: org.scalatest.exceptions.TestFailedException: task0.isDefined was true, but task1.isDefined was false
	at org.scalatest.Assertions.newAssertionFailedException(Assertions.scala:530)
	at org.scalatest.Assertions.newAssertionFailedException$(Assertions.scala:529)
	at org.scalatest.FunSuite.newAssertionFailedException(FunSuite.scala:1560)
	at org.scalatest.Assertions$AssertionsHelper.macroAssert(Assertions.scala:503)
	at org.apache.spark.scheduler.TaskSetManagerSuite.$anonfun$new$107(TaskSetManagerSuite.scala:1933)
	at org.scalatest.OutcomeOf.outcomeOf(OutcomeOf.scala:85)
	at org.scalatest.OutcomeOf.outcomeOf$(OutcomeOf.scala:83)
	at org.scalatest.OutcomeOf$.outcomeOf(OutcomeOf.scala:104)
	at org.scalatest.Transformer.apply(Transformer.scala:22)
	at org.scalatest.Transformer.apply(Transformer.scala:20)
	at org.scalatest.FunSuiteLike$$anon$1.apply(FunSuiteLike.scala:186)
	at org.apache.spark.SparkFunSuite.withFixture(SparkFunSuite.scala:149)
	at org.scalatest.FunSuiteLike.invokeWithFixture$1(FunSuiteLike.scala:184)
	at org.scalatest.FunSuiteLike.$anonfun$runTest$1(FunSuiteLike.scala:196)
	at org.scalatest.SuperEngine.runTestImpl(Engine.scala:286)
	at org.scalatest.FunSuiteLike.runTest(FunSuiteLike.scala:196)
	at org.scalatest.FunSuiteLike.runTest$(FunSuiteLike.scala:178)
	at org.apache.spark.SparkFunSuite.org$scalatest$BeforeAndAfterEach$$super$runTest(SparkFunSuite.scala:56)
	at org.scalatest.BeforeAndAfterEach.runTest(BeforeAndAfterEach.scala:221)
	at org.scalatest.BeforeAndAfterEach.runTest$(BeforeAndAfterEach.scala:214)
	at org.apache.spark.SparkFunSuite.runTest(SparkFunSuite.scala:56)
	at org.scalatest.FunSuiteLike.$anonfun$runTests$1(FunSuiteLike.scala:229)
	at org.scalatest.SuperEngine.$anonfun$runTestsInBranch$1(Engine.scala:393)
	at scala.collection.immutable.List.foreach(List.scala:392)
	at org.scalatest.SuperEngine.traverseSubNodes$1(Engine.scala:381)
	at org.scalatest.SuperEngine.runTestsInBranch(Engine.scala:376)
	at org.scalatest.SuperEngine.runTestsImpl(Engine.scala:458)
	at org.scalatest.FunSuiteLike.runTests(FunSuiteLike.scala:229)
	at org.scalatest.FunSuiteLike.runTests$(FunSuiteLike.scala:228)
	at org.scalatest.FunSuite.runTests(FunSuite.scala:1560)
	at org.scalatest.Suite.run(Suite.scala:1124)
	at org.scalatest.Suite.run$(Suite.scala:1106)
	at org.scalatest.FunSuite.org$scalatest$FunSuiteLike$$super$run(FunSuite.scala:1560)
	at org.scalatest.FunSuiteLike.$anonfun$run$1(FunSuiteLike.scala:233)
	at org.scalatest.SuperEngine.runImpl(Engine.scala:518)
	at org.scalatest.FunSuiteLike.run(FunSuiteLike.scala:233)
	at org.scalatest.FunSuiteLike.run$(FunSuiteLike.scala:232)
	at org.apache.spark.SparkFunSuite.org$scalatest$BeforeAndAfterAll$$super$run(SparkFunSuite.scala:56)
	at org.scalatest.BeforeAndAfterAll.liftedTree1$1(BeforeAndAfterAll.scala:213)
	at org.scalatest.BeforeAndAfterAll.run(BeforeAndAfterAll.scala:210)
	at org.scalatest.BeforeAndAfterAll.run$(BeforeAndAfterAll.scala:208)
	at org.apache.spark.SparkFunSuite.run(SparkFunSuite.scala:56)
	at org.scalatest.tools.Framework.org$scalatest$tools$Framework$$runSuite(Framework.scala:317)
	at org.scalatest.tools.Framework$ScalaTestTask.execute(Framework.scala:510)
	at sbt.ForkMain$Run$2.call(ForkMain.java:296)
	at sbt.ForkMain$Run$2.call(ForkMain.java:286)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
```

Doesn't look to be related, though I'm seeing this failure at the first time.

HeartSaVioR commented:
retest this, please

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/20924/
Test PASSed.

SparkQA commented:
**[Test build #116132 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/116132/testReport)** for PR 27085 at commit [`9a7a210`](https://github.com/apache/spark/commit/9a7a210d8b1214d7d9251bcf6b32a412e1d6e803).

SparkQA commented:
**[Test build #116132 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/116132/testReport)** for PR 27085 at commit [`9a7a210`](https://github.com/apache/spark/commit/9a7a210d8b1214d7d9251bcf6b32a412e1d6e803).
 * This patch **fails due to an unknown error code, -9**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `case class CompactionResult(code: CompactionResultCode.Value, compactIndex: Option[Long])`

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/116132/
Test FAILed.

HeartSaVioR commented:
retest this, please

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/20939/
Test PASSed.

SparkQA commented:
**[Test build #116148 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/116148/testReport)** for PR 27085 at commit [`9a7a210`](https://github.com/apache/spark/commit/9a7a210d8b1214d7d9251bcf6b32a412e1d6e803).

SparkQA commented:
**[Test build #116148 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/116148/testReport)** for PR 27085 at commit [`9a7a210`](https://github.com/apache/spark/commit/9a7a210d8b1214d7d9251bcf6b32a412e1d6e803).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `case class CompactionResult(code: CompactionResultCode.Value, compactIndex: Option[Long])`

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/116148/
Test FAILed.

HeartSaVioR commented:
Test failure on build 116148 is below:

```
org.apache.spark.deploy.StandaloneDynamicAllocationSuite.dynamic allocation default behavior
 Error Details
org.scalatest.exceptions.TestFailedException: StandaloneDynamicAllocationSuite.this.killAllExecutors(StandaloneDynamicAllocationSuite.this.sc) was false
 Stack Trace
sbt.ForkMain$ForkError: org.scalatest.exceptions.TestFailedException: StandaloneDynamicAllocationSuite.this.killAllExecutors(StandaloneDynamicAllocationSuite.this.sc) was false
	at org.scalatest.Assertions.newAssertionFailedException(Assertions.scala:530)
	at org.scalatest.Assertions.newAssertionFailedException$(Assertions.scala:529)
	at org.scalatest.FunSuite.newAssertionFailedException(FunSuite.scala:1560)
	at org.scalatest.Assertions$AssertionsHelper.macroAssert(Assertions.scala:503)
	at org.apache.spark.deploy.StandaloneDynamicAllocationSuite.$anonfun$new$1(StandaloneDynamicAllocationSuite.scala:126)
	at org.scalatest.OutcomeOf.outcomeOf(OutcomeOf.scala:85)
	at org.scalatest.OutcomeOf.outcomeOf$(OutcomeOf.scala:83)
	at org.scalatest.OutcomeOf$.outcomeOf(OutcomeOf.scala:104)
	at org.scalatest.Transformer.apply(Transformer.scala:22)
	at org.scalatest.Transformer.apply(Transformer.scala:20)
	at org.scalatest.FunSuiteLike$$anon$1.apply(FunSuiteLike.scala:186)
	at org.apache.spark.SparkFunSuite.withFixture(SparkFunSuite.scala:149)
	at org.scalatest.FunSuiteLike.invokeWithFixture$1(FunSuiteLike.scala:184)
	at org.scalatest.FunSuiteLike.$anonfun$runTest$1(FunSuiteLike.scala:196)
	at org.scalatest.SuperEngine.runTestImpl(Engine.scala:286)
	at org.scalatest.FunSuiteLike.runTest(FunSuiteLike.scala:196)
	at org.scalatest.FunSuiteLike.runTest$(FunSuiteLike.scala:178)
	at org.apache.spark.SparkFunSuite.org$scalatest$BeforeAndAfterEach$$super$runTest(SparkFunSuite.scala:56)
	at org.scalatest.BeforeAndAfterEach.runTest(BeforeAndAfterEach.scala:221)
	at org.scalatest.BeforeAndAfterEach.runTest$(BeforeAndAfterEach.scala:214)
	at org.apache.spark.SparkFunSuite.runTest(SparkFunSuite.scala:56)
	at org.scalatest.FunSuiteLike.$anonfun$runTests$1(FunSuiteLike.scala:229)
	at org.scalatest.SuperEngine.$anonfun$runTestsInBranch$1(Engine.scala:393)
	at scala.collection.immutable.List.foreach(List.scala:392)
	at org.scalatest.SuperEngine.traverseSubNodes$1(Engine.scala:381)
	at org.scalatest.SuperEngine.runTestsInBranch(Engine.scala:376)
	at org.scalatest.SuperEngine.runTestsImpl(Engine.scala:458)
	at org.scalatest.FunSuiteLike.runTests(FunSuiteLike.scala:229)
	at org.scalatest.FunSuiteLike.runTests$(FunSuiteLike.scala:228)
	at org.scalatest.FunSuite.runTests(FunSuite.scala:1560)
	at org.scalatest.Suite.run(Suite.scala:1124)
	at org.scalatest.Suite.run$(Suite.scala:1106)
	at org.scalatest.FunSuite.org$scalatest$FunSuiteLike$$super$run(FunSuite.scala:1560)
	at org.scalatest.FunSuiteLike.$anonfun$run$1(FunSuiteLike.scala:233)
	at org.scalatest.SuperEngine.runImpl(Engine.scala:518)
	at org.scalatest.FunSuiteLike.run(FunSuiteLike.scala:233)
	at org.scalatest.FunSuiteLike.run$(FunSuiteLike.scala:232)
	at org.apache.spark.SparkFunSuite.org$scalatest$BeforeAndAfterAll$$super$run(SparkFunSuite.scala:56)
	at org.scalatest.BeforeAndAfterAll.liftedTree1$1(BeforeAndAfterAll.scala:213)
	at org.scalatest.BeforeAndAfterAll.run(BeforeAndAfterAll.scala:210)
	at org.scalatest.BeforeAndAfterAll.run$(BeforeAndAfterAll.scala:208)
	at org.apache.spark.SparkFunSuite.run(SparkFunSuite.scala:56)
	at org.scalatest.tools.Framework.org$scalatest$tools$Framework$$runSuite(Framework.scala:317)
	at org.scalatest.tools.Framework$ScalaTestTask.execute(Framework.scala:510)
	at sbt.ForkMain$Run$2.call(ForkMain.java:296)
	at sbt.ForkMain$Run$2.call(ForkMain.java:286)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
```

Also doesn't look to be related.

HeartSaVioR commented:
retest this, please

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/20951/
Test PASSed.

SparkQA commented:
**[Test build #116159 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/116159/testReport)** for PR 27085 at commit [`9a7a210`](https://github.com/apache/spark/commit/9a7a210d8b1214d7d9251bcf6b32a412e1d6e803).

SparkQA commented:
**[Test build #116159 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/116159/testReport)** for PR 27085 at commit [`9a7a210`](https://github.com/apache/spark/commit/9a7a210d8b1214d7d9251bcf6b32a412e1d6e803).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `case class CompactionResult(code: CompactionResultCode.Value, compactIndex: Option[Long])`

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/116159/
Test FAILed.

HeartSaVioR commented:
Test failure on build 116159 is below:

```
> org.apache.spark.deploy.master.MasterSuite.master/worker web ui available

org.scalatest.exceptions.TestFailedDueToTimeoutException: The code passed to eventually never returned normally. Attempted 45 times over 5.0144778059999995 seconds. Last failure message: Server returned HTTP response code: 500 for URL: http://amp-jenkins-worker-06.amp:35696/json/.

> org.apache.spark.deploy.master.MasterSuite.master/worker web ui available with reverseProxy

org.scalatest.exceptions.TestFailedDueToTimeoutException: The code passed to eventually never returned normally. Attempted 46 times over 5.059962303 seconds. Last failure message: Server returned HTTP response code: 500 for URL: http://localhost:45829/proxy/worker-20200106030855-amp-jenkins-worker-06.amp-36774/json/.

> org.apache.spark.scheduler.TaskSetManagerSuite.reset

org.scalatest.exceptions.TestFailedException: task0.isDefined was true, but task1.isDefined was false
```

There's a PR #27010 to address above two test failures. For the last test failure, we may have to file a new one if there's no JIRA issue for this.

HeartSaVioR commented:
retest this, please

SparkQA commented:
**[Test build #116171 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/116171/testReport)** for PR 27085 at commit [`9a7a210`](https://github.com/apache/spark/commit/9a7a210d8b1214d7d9251bcf6b32a412e1d6e803).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/20964/
Test PASSed.

HeartSaVioR commented:
FYI, filed SPARK-30438 for `Flaky test: org.apache.spark.deploy.StandaloneDynamicAllocationSuite."dynamic allocation default behavior"` and filed SPARK-30440 for `Flaky test: org.apache.spark.scheduler.TaskSetManagerSuite.reset`.

SparkQA commented:
**[Test build #116171 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/116171/testReport)** for PR 27085 at commit [`9a7a210`](https://github.com/apache/spark/commit/9a7a210d8b1214d7d9251bcf6b32a412e1d6e803).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `case class CompactionResult(code: CompactionResultCode.Value, compactIndex: Option[Long])`

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/116171/
Test PASSed.

HeartSaVioR committed:
Reflect review comments

SparkQA commented:
**[Test build #116211 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/116211/testReport)** for PR 27085 at commit [`be77ce1`](https://github.com/apache/spark/commit/be77ce1343072e3d77bd3027859707c844804b9c).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/21003/
Test PASSed.

SparkQA commented:
**[Test build #116211 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/116211/testReport)** for PR 27085 at commit [`be77ce1`](https://github.com/apache/spark/commit/be77ce1343072e3d77bd3027859707c844804b9c).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/116211/
Test PASSed.

vanzin commented:
> Btw, would it be better to file a new issue for each follow-up PR

I'd create new tasks for those.

vanzin commented on core/src/main/scala/org/apache/spark/deploy/history/BasicEventFilterBuilder.scala:
```diff
@@ -0,0 +1,178 @@
+  }
+
+  override def onTaskStart(taskStart: SparkListenerTaskStart): Unit = {
+    _stageToTasks.get(taskStart.stageId) match {
```
use `foreach` instead of `match`

vanzin commented on core/src/main/scala/org/apache/spark/deploy/history/BasicEventFilterBuilder.scala:
```diff
@@ -0,0 +1,178 @@
+  override def onTaskStart(taskStart: SparkListenerTaskStart): Unit = {
+    _stageToTasks.get(taskStart.stageId) match {
+      case Some(tasks) =>
+        totalTasks += 1
```
Shouldn't you increment totalTasks always? Since this task event will be filtered (and thus it counts towards the statistics about how much compaction will save).

HeartSaVioR replied:
You're right. My bad. I felt we want to ignore the event but yes it should be counted in any way.

vanzin commented on core/src/main/scala/org/apache/spark/deploy/history/EventLogFileCompactor.scala:
```diff
@@ -0,0 +1,224 @@
+  }
+
+  /**
+   * This class rewrites the event log files into one compact file: the compact file will only
```
s/class/method

vanzin commented on core/src/main/scala/org/apache/spark/deploy/history/EventLogFileCompactor.scala:
```diff
@@ -0,0 +1,224 @@
+ * @param compactIndex The index of compact file if the compaction is successful.
+ *                     Otherwise it will be None.
+ */
+case class CompactionResult(code: CompactionResultCode.Value, compactIndex: Option[Long])
```
can probably be private

HeartSaVioR replied:
This is used as the return type of `compact`, hence have to follow the scope of the method - it's public.

HeartSaVioR replied:
As we marked the scope of EventFilter as `private[spark]`, I'll change EventLogFileCompactor to `private[spark]` and change CompactionResult/CompactionResultCode to `private[spark]` as well. 

Maybe we could make it tighter (`private[history]` or `private[deploy]`) - please let me know if it's preferred. Thanks!

vanzin replied:
Hmm. I guess this package is not considered a public API (as nothing under `deploy` is), so in those cases I've been leaning towards just leaving the class as public. `private` makes sense if the class doesn't exist the scope of the package, which is not the case here.

HeartSaVioR replied:
OK will leave them as public. Thanks.

vanzin commented on core/src/main/scala/org/apache/spark/deploy/history/EventLogFileCompactor.scala:
```diff
@@ -0,0 +1,224 @@
+ */
+case class CompactionResult(code: CompactionResultCode.Value, compactIndex: Option[Long])
+
+object CompactionResultCode extends Enumeration {
```
can probably be private

vanzin commented on core/src/test/scala/org/apache/spark/deploy/history/EventLogFileCompactorSuite.scala:
```diff
@@ -0,0 +1,326 @@
+      // filterApplicationEnd: Some(true) & Some(true) => filter in
+      expectedLines += writeEventToWriter(writer, SparkListenerApplicationEnd(0))
+
+      // filterBlockManagerAdded: Some(true) & Some(false) => filter out
```
OOC, can you give an example of this behavior in a real log file?

It seems a bit weird to filter out an event when at least one filter has decided the event should be kept (maybe because it's needed for something?).

HeartSaVioR replied:
Yeah... I agree that's somewhat unintuitive and not working same as normal chain of filters.

It's due to the difference of acceptance of jobs between event filter in CORE and event filter in SQL. In SQL event filter, we don't drop finished job if the job relates to the SQL execution and the SQL execution is live.

In BasicEventFilter there's no way to indicate SQL execution event so it rejects all finished jobs, but SQLEventFilter will work a bit differently as I described, so there will be some jobs which BasicEventFilter decides to reject whereas SQLEventFilter decides to accept - we should accept these jobs in this case.

That was the main reason I had to introduce None along with true / false.

vanzin replied:
> here will be some jobs which BasicEventFilter decides to reject whereas SQLEventFilter decides to accept - we should accept these jobs in this case

So in that case the basic filter will say `Some(false)`, and SQL will say `Some(true)`, and the event should be accepted, right? Isn't that the opposite what this comment is saying?

HeartSaVioR replied:
Oh wait you're right. Looks like something was messed up when introducing None. Let me fix it immediately.

vanzin reviewed (commented):
A few minor things but I think this is looking ok now.

HeartSaVioR committed:
Reflect review comments

SparkQA commented:
**[Test build #116328 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/116328/testReport)** for PR 27085 at commit [`29ba0f2`](https://github.com/apache/spark/commit/29ba0f2bc937d459cea434061f921050b68b8c73).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/21118/
Test PASSed.

HeartSaVioR committed:
Fixed my bad

SparkQA commented:
**[Test build #116330 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/116330/testReport)** for PR 27085 at commit [`163bda0`](https://github.com/apache/spark/commit/163bda018d054a0d41e983e6c448ce3b3e746d35).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/21120/
Test PASSed.

SparkQA commented:
**[Test build #116328 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/116328/testReport)** for PR 27085 at commit [`29ba0f2`](https://github.com/apache/spark/commit/29ba0f2bc937d459cea434061f921050b68b8c73).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/116328/
Test PASSed.

SparkQA commented:
**[Test build #116330 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/116330/testReport)** for PR 27085 at commit [`163bda0`](https://github.com/apache/spark/commit/163bda018d054a0d41e983e6c448ce3b3e746d35).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `class EventLogFileCompactor(`
  * `case class CompactionResult(code: CompactionResultCode.Value, compactIndex: Option[Long])`

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/116330/
Test PASSed.

vanzin commented:
I don't think I have more comments; will leave this here a little bit to see if anyone shows up.

Only thing that crossed my mind is that maybe it might be useful to separate compaction from cleaning; you can have compaction to speed up generation of the history UI, but keep all the log files if you need to debug something in detail.

HeartSaVioR commented:
Thanks for the reviews, as always.

Given we decide to break down PR into multiple incremental PRs for faster reviewing, I'd also like to kindly ask about classifying items into "mandatory" and "good to have", and deal with "mandatory" items with fast iterations.

The main goal for me is getting all of parts be merged within feature freeze date. I guess tentative date is set to Jan 31, but there's lunar new year in later this month which is one of most big holidays in S.Korea (China as well), so may need to exclude around a week (I'll try to stick with this though), and then there's not much time. Hopefully there're relatively smaller PRs left for reviewing, but integration part may also need to have enough time to review as the feature starts to make effect after that.

Details/improvements could be addressed afterwards once the feature is in, and then the feature is no longer be tied with only me, anyone should be able to deal with these items.

HeartSaVioR commented:
Enumerated items in JIRA issue which were mentioned but didn't addressed as they weren't considered as mandatory.

https://issues.apache.org/jira/browse/SPARK-28594?focusedCommentId=17012562&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-17012562

We can add items whenever we find more, and create sub-task if we feel something as mandatory, or someone wants to collaborate.

vanzin commented:
Ok, merging this one to master. On to the next.

HeartSaVioR commented:
Thanks for reviewing and merging! I'll rebase the next PR and ping you.

HyukjinKwon commented on core/src/main/resources/META-INF/services/org.apache.spark.deploy.history.EventFilterBuilder:
```diff
@@ -0,0 +1 @@
+org.apache.spark.deploy.history.BasicEventFilterBuilder
```
`EventFilterBuilder` is private in Spark. Do you mind if I ask why we use service loader?

HeartSaVioR replied:
core module cannot refer SQL/core module which also has EventFilterBuilder.

HyukjinKwon replied:
I see. I think that's possible via simply using reflection which I think is easier to read the codes. I think we're already doing this in few places such as `FileCommitProtocol.instantiate`

Seems a bit odds to use service loader for internal classes.

HeartSaVioR replied:
https://github.com/apache/spark/blob/a5c7090ffafbfb4a8f312251b9ce8f25f71497cb/core/src/main/scala/org/apache/spark/deploy/history/FsHistoryProvider.scala#L1226-L1228

That is the way how FsHistoryProvider already deals with loading class in SQL core module.

HyukjinKwon replied:
Okay, thanks. at least it's consistent.

HyukjinKwon commented on core/src/main/scala/org/apache/spark/internal/config/package.scala:
```diff
@@ -195,6 +195,24 @@ package object config {
+
+  private[spark] val EVENT_LOG_COMPACTION_SCORE_THRESHOLD =
+    ConfigBuilder("spark.eventLog.rolling.compaction.score.threshold")
+      .internal()
```
I think we should have added some docs here too.

HeartSaVioR replied:
It was to ease of test for compaction but after reviewing it's no longer need to be a configuration.
I'll just remove this and convert it to be constant unless we figure out needs to adjust manually.

HeartSaVioR replied:
Just realized there's test to modify the value to ensure compaction doesn't skip with low value. I'll retain the configuration and add some doc.

HyukjinKwon commented on core/src/main/scala/org/apache/spark/internal/config/package.scala:
```diff
@@ -195,6 +195,24 @@ package object config {
+        "the overall size of event log files.")
+      .intConf
+      .checkValue(_ > 0, "Max event log files to retain should be higher than 0.")
+      .createWithDefault(Integer.MAX_VALUE)
```
Why didn't we make it optional, or defaults to -1 to express "all event log files will be retained"?

HeartSaVioR replied:
I think this default value "naturally" fits the semantic of the configuration even we don't indicate Integer.MAX_VALUE as special value. No need to care the configuration specially. This approach seems to be already used for couple of configurations.
